### PR TITLE
esn#11 - Setting up tests for calendar

### PR DESCRIPTION
**Pull request by ppppbn**
Partially resolve https://github.com/OpenPaaS-Suite/esn/issues/11

### CODE REVIEW

File: src/esn.calendar.libs/app/services/calendar-api.spec.js [hunk 1]
```diff
@@ -29,9 +29,9 @@ describe('The calendar module apis', function() {
   }
 
   beforeEach(function() {
-    module('esn.calendar.libs');
+    angular.mock.module('esn.calendar.libs');
```
Comment (chamerling) [inline src/esn.calendar.libs/app/services/calendar-api.spec.js:32]: Instead of having to do this everywhere, you can tell webpack to do it for you then you will not have to change all the files

File: src/index.test.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+window.jstz = require('esn-frontend-common-libs/src/frontend/components/jstzdetect/jstz.js');
+window.jQuery = require('jquery/dist/jquery.js');
+window.$ = window.jQuery;
+window.$q = require('q/q.js');
```
Comment (chamerling) [inline src/index.test.js:4]: Can't we rely on webpack for this?

Comment (ppppbn) [reply src/index.test.js:4]: I tried with webpack but some throw errors when using with `$window` of AngularJS. Therefore I have to put them here. Will try to refactor this once every test run smoothly.

File: package.json [hunk 3]
```diff
@@ -4,26 +4,33 @@
...
     "serve:prod": "webpack-dev-server --env=prod",
     "build:dev": "webpack",
-    "build:prod": "webpack --env=prod"
+    "build:prod": "webpack --env=prod",
+    "build:test": "webpack --env=test"
   },
   "author": "",
   "license": "AGPLv3",
   "dependencies": {
+    "@linagora/ical.js": "1.2.4",
```
Comment (tuanlc) [inline package.json:17]: Is this used only for test? If yes, it should belong to `devDependencies`

File: package.json [hunk 4]
```diff
@@ -4,26 +4,33 @@
...
     "esn-frontend-common-libs": "github:openpaas-suite/esn-frontend-common-libs#main",
     "esn-frontend-mailto-handler": "github:openpaas-suite/esn-frontend-mailto-handler#main",
     "esn-frontend-videoconference-calendar": "github:openpaas-suite/esn-frontend-videoconference-calendar#main",
-    "fullcalendar": "3.9.0"
+    "fullcalendar": "3.9.0",
+    "q": "^1.5.1",
+    "restangular": "^1.6.1"
```
Comment (tuanlc) [inline package.json:23]: Why do we need them? For test also?

File: src/esn.calendar.libs/app/components/attendee/list/attendee-list.controller.spec.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ describe('The CalAttendeeListController controller', function() {
   var calAttendeeService, context, $controller, $q, $scope, $rootScope;
 
   beforeEach(function() {
-    module('jadeTemplates');
+    ;
```
Comment (tuanlc) [inline src/esn.calendar.libs/app/components/attendee/list/attendee-list.controller.spec.js:11]: :100:

File: src/linagora.esn.resource/app/components/resource-form-modal/create/resource-form-create.controller.spec.js [hunk 6]
```diff
@@ -8,7 +8,7 @@ describe('The ESNResourceFormCreateController controller', function() {
   var $q, $state, $controller, $rootScope, $scope, context, resource, esnResourceAPIClient, typeMock, sessionMock, asyncAction, resourceType;
 
   beforeEach(function() {
-    module('jadeTemplates');
+    ;
```
Comment (tuanlc) [inline src/linagora.esn.resource/app/components/resource-form-modal/create/resource-form-create.controller.spec.js:11]: :metal:

File: src/linagora.esn.resource/app/components/resource-form-modal/update/resource-form-update.controller.spec.js [hunk 7]
```diff
@@ -8,7 +8,7 @@ describe('The ESNResourceFormUpdateController controller', function() {
   var $q, $state, $controller, $rootScope, resource, resourceUpdated, esnResourceAPIClient, sessionMock, asyncAction, resourceType;
 
   beforeEach(function() {
-    module('jadeTemplates');
+    ;
```
Comment (tuanlc) [inline src/linagora.esn.resource/app/components/resource-form-modal/update/resource-form-update.controller.spec.js:11]: :metal:

File: src/linagora.esn.resource/app/components/resource-icon-picker/resource-icon-picker.controller.spec.js [hunk 8]
```diff
@@ -9,7 +9,7 @@ describe('The esnResourceIconPicker controller', function() {
   var ctrl;
 
   beforeEach(function() {
-    module('jadeTemplates');
+    ;
```
Comment (tuanlc) [inline src/linagora.esn.resource/app/components/resource-icon-picker/resource-icon-picker.controller.spec.js:12]: :metal:

File: test/config/db.json [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (tuanlc) [inline test/config/db.json:1]: Do we need this since the test only for frontend test

File: test/config/default.json [hunk 10]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (tuanlc) [inline test/config/default.json:1]: same for this file, no need for frontend test

File: test/config/mocks/i18n-loader.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
+(function(angular) {
```
Comment (tuanlc) [inline test/config/mocks/i18n-loader.js:1]: No need IIFE

File: test/config/mocks/i18n-loader.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  'use strict';
+
+  angular.module('esn.i18n')
+    .factory('esnI18nLoader', esnI18nLoader);
```
Comment (tuanlc) [inline test/config/mocks/i18n-loader.js:5]: Why do not mock the module like other mocked modules?

File: test/config/mocks/modules.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+angular.module('esn.calendar.event-consultation', []);
+angular.module('ngCookies', []);
+angular.module('pascalprecht.translate', [])
+  .service('$translate', function() {
+    return {
+      preferredLanguage: function() { return ''; },
+      use: function() {},
+      instant: function() {}
+    };
+  });
```
Comment (tuanlc) [inline test/config/mocks/modules.js:14]: add blank line after

File: test/config/mocks/reset-dynamic-directive-injections.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Angular providers do not reset between each test, only service instance do.
+ */
+(function(angular) {
```
Comment (tuanlc) [inline test/config/mocks/reset-dynamic-directive-injections.js:4]: remove IIFE

File: test/config/servers-conf.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
+'use strict';
```
Comment (tuanlc) [inline test/config/servers-conf.js:1]: Frontend test does not need this file also

File: webpack.test.js [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            loader: 'pug-loader',
+            options: pugLoaderOptions
+          },
+        ],
+      },
+      { 
+        test: /\.(js|jsx)$/,
+        use: [
+          {
+            loader: "babel-loader"
```
Comment (tuanlc) [inline webpack.test.js:139]: Question for all: Do we really need `babel` to run tests? I assume that the chosen browser headless support ES6

@MichaelBailly @chamerling @huy-ta @piavgh @ppppbn

Comment (ppppbn) [reply webpack.test.js:139]: Tested without `babel` and the performance improved a lot. I think we can get rid of `babel`

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 17]
```diff
@@ -357,19 +377,20 @@ describe('The CalEventFormController controller', function() {
...
-        this.scope.attendees.users[0].freeBusy = 'busy';
-        this.scope.submit();
-        this.rootScope.$digest();
-
-        expect(calEventServiceMock.createEvent).to.not.have.been.called;
-        expect(calEventFreeBusyConfirmationModalService).to.have.been.called;
+        this.initController().then(function() {
+          self.scope.attendees.users[0].freeBusy = 'busy';
+          self.scope.submit();
+          self.rootScope.$digest();
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:383]: In the test above, you removed this line, but here you did not. Is there a difference?

File: src/esn.calendar.libs/app/components/event-recurrence-edition/event-recurrence-edition.spec.js [hunk 18]
```diff
@@ -45,115 +51,125 @@ describe('The event-recurrence-edition component', function() {
...
-      this.eleScope.selectEndRadioButton(2);
-      var radio = angular.element(element).find('input[name="inlineRadioEndOptions"]')[2];
-
-      expect(radio.checked).to.be.true;
+      this.initDirective(this.$scope).then(function(element) {
+        self.eleScope.selectEndRadioButton(2);
+        var radio = angular.element(element).find('input[name="inlineRadioEndOptions"]')[2];
+  
+        expect(radio.checked).to.be.true;
+      }).then(done)
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event-recurrence-edition/event-recurrence-edition.spec.js:149]:
```suggestion
        expect(radio.checked).to.be.true;
        done();
      })
```

File: src/esn.calendar.libs/app/components/event-recurrence-edition/event-recurrence-edition.spec.js [hunk 19]
```diff
@@ -45,115 +51,125 @@ describe('The event-recurrence-edition component', function() {
...
+        expect(self.eleScope.vm.days).to.shallowDeepEqual([
+          {value: 'MO', selected: false},
+          {value: 'TU', selected: false},
+          {value: 'WE', selected: false},
+          {value: 'TH', selected: false},
+          {value: 'FR', selected: false},
+          {value: 'SA', selected: true},
+          {value: 'SU', selected: true}
+        ]);
+      }).then(done);
```
Comment (tuanlc) [inline src/esn.calendar.libs/app/components/event-recurrence-edition/event-recurrence-edition.spec.js:83]: Please  `catch` all the errors in this PR to have clear logs

```suggestion
      }).then(done)
      .catch(err => done(err || new Error('should resolve'));
```

File: src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js [hunk 20]
```diff
@@ -383,32 +395,41 @@ describe('The calFreebusyService service', function() {
...
 
-      it('should set freebusy to unknow for internal users when freebusy status is not found', function() {
+      it('should set freebusy to unknow for internal users when freebusy status is not found', function(done) {
         getBulkFreebusyStatusStub.returns($q.when({}));
         calFreebusyService.setBulkFreeBusyStatus(attendees, start, end, [event]);
         $rootScope.$digest();
 
-        expect(attendees[0].freeBusy).to.equal(CAL_FREEBUSY.UNKNOWN);
-        expect(attendees[1].freeBusy).to.equal(CAL_FREEBUSY.UNKNOWN);
+        setTimeout(() => {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js:425]: Why do we need `setTimeout` here? Shouldn't `$rootScope.$digest()` do the job already?

File: src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js [hunk 21]
```diff
@@ -24,14 +24,15 @@ describe('the calendarAttendeeService', function() {
...
+    angular.mock.inject(function(_$rootScope_, _CAL_ATTENDEE_OBJECT_TYPE_, _CAL_ICAL_, _$q_) {
       $rootScope = _$rootScope_;
       CAL_ATTENDEE_OBJECT_TYPE = _CAL_ATTENDEE_OBJECT_TYPE_;
       CAL_ICAL = _CAL_ICAL_;
+      $q = _$q_;
     });
   });
 
-  beforeEach(angular.mock.inject(function(_calendarAttendeeService_) {
+  beforeEach(inject(function(_calendarAttendeeService_) {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js:35]: In some other places you added `angular.mock`, but in this you removed it intentionally?

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration.controller.js [hunk 22]
```diff
@@ -137,11 +140,12 @@ function calendarConfigurationController(
 
               return calendarAPI.modifyPublicRights(self.calendarHomeId, self.calendar.id, { public_right: self.publicSelection });
             default:
-              return $q.when();
+              return;
           }
         })
         .then(function() {
-          notificationFactory.weakInfo('New calendar -', esnI18nService.translate('%s has been created', { name: self.calendar.name }));
+          notificationFactory.weakInfo('New calendar -', esnI18nService.translate('%s has been created', self.calendar.name));
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration.controller.js:147]: Why? This would break the translation.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration.controller.js [hunk 23]
```diff
@@ -181,8 +185,8 @@ function calendarConfigurationController(
         updateActions.push(calendarAPI.modifyPublicRights(self.calendarHomeId, self.calendar.id, { public_right: self.publicSelection }));
       }
 
-      $q.all(updateActions).then(function() {
-        notificationFactory.weakInfo('Calendar -', esnI18nService.translate('%s has been modified.', { name: self.calendar.name }));
+      return Promise.all(updateActions).then(function() {
+        notificationFactory.weakInfo('Calendar -', esnI18nService.translate('%s has been modified.', self.calendar.name));
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration.controller.js:189]: Why? This would break the translation.

File: src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.directive.spec.js [hunk 24]
```diff
@@ -15,7 +15,7 @@ describe('calendarView directive', function() {
       }
     };
 
-    module('jadeTemplates');
+    ;
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.directive.spec.js:18]: bleh

File: src/linagora.esn.calendar/app/calendar/calendar.controller.spec.js [hunk 25]
```diff
@@ -9,7 +9,7 @@ describe('The CalCalendarController controller', function() {
   var calFullUiConfiguration, $q, $controller, $rootScope, $scope, context, businessHours;
 
   beforeEach(function() {
-    module('jadeTemplates');
+    // ;
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar/calendar.controller.spec.js:12]: bruhhh

File: src/linagora.esn.calendar/app/components/calendar-color-picker/calendar-color-picker.spec.js [hunk 26]
```diff
@@ -9,7 +9,7 @@ describe('The calendar color picker component', function() {
 
   beforeEach(function() {
     self = this;
-    module('jadeTemplates');
+    ;
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/components/calendar-color-picker/calendar-color-picker.spec.js:12]: pffttt

File: src/linagora.esn.calendar/app/components/calendars-list/calendars-list.controller.js [hunk 27]
```diff
@@ -31,12 +31,12 @@ const _ = require('lodash');
...
       self.toggleCalendar = calendarVisibilityService.toggle;
 
-      self.activate();
+      return self.activate();
     }
 
     function activate() {
-      $q
-        .all(listCalendars(), getHiddenCalendars())
+      return Promise
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/components/calendars-list/calendars-list.controller.js:38]: Is there a reanson why sometimes `$q` needs to be changed to `Promise`, but not always?

File: src/linagora.esn.calendar/app/components/mini-calendar/mini-calendar.controller.spec.js [hunk 28]
```diff
@@ -138,7 +138,7 @@ describe('The miniCalendarController controller', function() {
   }));
 
   afterEach(function() {
-    $scope.$destroy();
+    $scope && $scope.destroy && $scope.$destroy();
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/components/mini-calendar/mini-calendar.controller.spec.js:141]: Shouldn't it be:

```suggestion
    $scope && typeof $scope.$destroy === 'function' && $scope.$destroy();
```

By the way why do we need this check?

File: src/linagora.esn.calendar/app/core/date-to-moment.directive.spec.js [hunk 29]
```diff
@@ -6,7 +6,7 @@ var expect = chai.expect;
 
 describe('The calDateToMoment directive', function() {
   beforeEach(function() {
-    module('jadeTemplates');
+    ;
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/core/date-to-moment.directive.spec.js:9]: :kewk:

File: src/linagora.esn.calendar/app/core/friendlify-end-date.directive.spec.js [hunk 30]
```diff
@@ -7,7 +7,7 @@ var expect = chai.expect;
 describe('The calFriendlifyEndDate directive', function() {
 
   beforeEach(function() {
-    module('jadeTemplates');
+    ;
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/core/friendlify-end-date.directive.spec.js:10]: lolololol

File: src/esn.calendar.libs/app/components/event-recurrence-edition/event-recurrence-edition.spec.js (hunk 18 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/components/event-recurrence-edition/event-recurrence-edition.spec.js:149]: Updated. Writing another then to make it easier to write catch also. Please see the comment below from @tuanlc , I have resolved that.

File: src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js (hunk 20 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js:425]: Normally yes. However in many cases, `$rootScope.$digest()` does not wait for the promise to be fulfilled. Therefore we have to use some `setTimeout` to ensure our test cases run after the logic code finished.

If you have any better idea please let me know. Thanks!

File: src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js (hunk 21 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js:35]: For some cases like this an error was thrown: `angular.mock.inject` is not a function. Using `inject` makes it work fine.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration.controller.js (hunk 22 above)
Comment (ppppbn) [reply src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration.controller.js:147]: My bad :<

File: src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js (hunk 21 above)
Comment (huy-ta) [reply src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js:35]: Mh... weird. Why is it inconsistent like that? There are some other `beforeEach` blocks that you use `angular.mock.inject` without errors.

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js (hunk 17 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js]: Nothing, just forgot to remove it :(

File: src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js (hunk 20 above)
Comment (huy-ta) [reply src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js:425]: It's weird that the previous code worked somehow :trollface:.

File: src/linagora.esn.calendar/app/components/calendars-list/calendars-list.controller.js (hunk 27 above)
Comment (ppppbn) [reply src/linagora.esn.calendar/app/components/calendars-list/calendars-list.controller.js:38]: Sometimes $q creates a promise that cannot be triggered, both normal Js way and the $rootScope.$digest() way.
So to make it safe and easy to predict, I use Promise in some hard cases.

Comment (huy-ta) [reply src/linagora.esn.calendar/app/components/calendars-list/calendars-list.controller.js:38]: Inconsistency can be a problem here. We need to either use only `$q` or only `Promise`, mixing both of them can be confusing. Maybe we can create another issue to handle that problem.

File: src/linagora.esn.calendar/app/components/mini-calendar/mini-calendar.controller.spec.js (hunk 28 above)
Comment (ppppbn) [reply src/linagora.esn.calendar/app/components/mini-calendar/mini-calendar.controller.spec.js]: Thank you. This check was implemented because some of the test cases failed to execute $scope.$destroy(). I believe they cannot find `$scope` since it has been removed / destroyed at the end of the test.

File: babel.config.json [hunk 31]
```diff
@@ -4,5 +4,14 @@
...
     ["angularjs-annotate", { "es6": true, "dynamicImport": true, "explicitOnly": false }]
+  ],
+  "ignore": [
+    "node_modules/esn-frontend-common-libs/src/frontend/components/ical.js/build/*.js",
+    "node_modules/esn-frontend-common-libs/node_modules/restangular/**/*.js",
+    "node_modules/restangular/**/*.js",
+    "node_modules/@linagora/ical.js/**/*.js",
+    "node_modules/angular/angular.js",
+    "node_modules/fullcalendar/dist/fullcalendar.js",
+    "node_modules/angular-material/angular-material.js"
```
Comment (huy-ta) [inline babel.config.json:15]: Do we need this now that we don't use Babel for tests?

File: src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js (hunk 21 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js:35]: This one `inject` was originally used as a wrapper function and returns a function, maybe that's why. Will try to refactor this soon. I'm gonna left this conversation unresolved to remind myself.

File: babel.config.json (hunk 31 above)
Comment (ppppbn) [reply babel.config.json:15]: No we don't, just forgot to remove this. Thanks!

File: test/config/mocks/i18n-loader.js [hunk 32]
```diff
@@ -0,0 +1,26 @@
+'use strict';
+
+angular.module('esn.i18n')
+  .factory('esnI18nLoader', esnI18nLoader);
+
+  function esnI18nLoader($q) {
```
Comment (huy-ta) [inline test/config/mocks/i18n-loader.js:6]: Wrong indentation.

File: src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js [hunk 33]
```diff
@@ -45,6 +47,8 @@ describe('The calFreebusyService service', function() {
   describe('The listFreebusy fn', function() {
 
     it('should list freebusy infos', function(done) {
+      var $qTemp = window.$q;
+      window.$q = $qLite;
```
Comment (tuanlc) [inline src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js:51]: This deserves an explanation

File: src/esn.calendar.libs/app/services/calendar-visibility.spec.js [hunk 34]
```diff
@@ -48,90 +48,87 @@ describe('The calendarVisibilityService', function() {
...
+      this.calendarVisibilityService.isHidden(cal).then(thenSpy).then(function() {
+        expect(thenSpy).to.have.been.calledWith(false);
+        thenSpy.reset();
+
+        self.calendarVisibilityService.toggle(cal).then(function() {
+          self.calendarVisibilityService.isHidden(cal).then(thenSpy).then(function() {
+            expect(thenSpy).to.have.been.calledWith(true);
+          });
+        }).then(done).catch(done);
+      });
```
Comment (tuanlc) [inline src/esn.calendar.libs/app/services/calendar-visibility.spec.js:131]: `catch` block should catch the outermost promise block

File: webpack.test.js [hunk 35]
```diff
@@ -0,0 +1,137 @@
...
+const pugLoaderOptions = {
+  root: `${__dirname}/node_modules/esn-frontend-common-libs/src/frontend/views`
+};
+const lodashPath = path.resolve(__dirname, 'node_modules', 'lodash', 'dist', 'lodash.js');
+const i18nLoaderMockPath = path.resolve(__dirname, 'test', 'config', 'mocks', 'i18n-loader.js');
+const icalPath = path.resolve(__dirname, 'node_modules', 'esn-frontend-common-libs', 'src', 'frontend', 'components', 'ical.js', 'build', 'ical.min.js');
+
+const BASE_HREF = process.env.BASE_HREF || '/';
+
+module.exports = {
```
Comment (huy-ta) [inline webpack.test.js:20]: Try using `webpack-merge`.

File: test/config/karma.conf.js [hunk 36]
```diff
@@ -0,0 +1,96 @@
...
+    ],
+    frameworks: ['mocha', 'sinon-chai'],
+    colors: true,
+    singleRun: singleRun,
+    autoWatch: true,
+    browsers: ['FirefoxHeadless'/*,'ChromeHeadless'*/],
+
+    customLaunchers: {
+      FirefoxHeadless: {base: 'Firefox', flags: ['--headless']},
+      ChromeHeadless: {base: 'Chrome', flags: ['--headless', '--disable-gpu']},
```
Comment (huy-ta) [inline test/config/karma.conf.js:37]:
```suggestion
      FirefoxHeadless: { base: 'Firefox', flags: ['--headless'] },
      ChromeHeadless: { base: 'Chrome', flags: ['--headless', '--disable-gpu'] },
```

File: src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js (hunk 20 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/components/freebusy/freebusy.service.spec.js:425]: :trollface:

File: test/config/karma.conf.js [hunk 37]
```diff
@@ -0,0 +1,96 @@
...
+      'karma-mocha',
+      'karma-webpack',
+      'karma-coverage',
+      'karma-spec-reporter',
+      'karma-sinon-chai',
+      '@linagora/karma-ng-jade2module-preprocessor',
+      'karma-rawfixtures-preprocessor'
+    ],
+
+    coverageReporter: {type: 'text', dir: '/tmp'},
```
Comment (huy-ta) [inline test/config/karma.conf.js:72]:
```suggestion
    coverageReporter: { type: 'text', dir: '/tmp' },
```

File: src/esn.calendar.libs/app/services/fullcalendar/calendar-configuration.service.spec.js [hunk 38]
```diff
@@ -100,7 +105,7 @@ describe('The calFullUiConfiguration service', function() {
           expect(uiConfiguration).to.shallowDeepEqual(CAL_UI_CONFIG);
           expect(esnUserConfigurationService.get).to.have.been.calledOnce;
           expect(esnUserConfigurationService.get).to.have.been.calledWith(moduleConfiguration, moduleName);
-
+          window.$q = $qTemp;
```
Comment (tuanlc) [inline src/esn.calendar.libs/app/services/fullcalendar/calendar-configuration.service.spec.js:108]: instead of doing this for every test case, you can use `beforeEach` to setup & `afterEach` for reset

And more, for the thing like this, it deserves an explanation, better if you have a common function with the good name, even jsdoc for it.

File: src/esn.calendar.libs/app/services/fullcalendar/calendar-configuration.service.spec.js [hunk 39]
```diff
@@ -5,22 +5,24 @@
 var expect = chai.expect;
 
 describe('The calFullUiConfiguration service', function() {
-  var $q,
-  $httpBackend,
+  var $httpBackend,
```
Comment (tuanlc) [inline src/esn.calendar.libs/app/services/fullcalendar/calendar-configuration.service.spec.js:8]: Why do you need this?

File: src/esn.calendar.libs/app/services/shells/calendar-right-shell.spec.js [hunk 40]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-/* global chai, __FIXTURES__, _: false */
+/* global chai, __FIXTURES__, _: true */
```
Comment (tuanlc) [inline src/esn.calendar.libs/app/services/shells/calendar-right-shell.spec.js:3]: Why?

File: webpack.test.js (hunk 35 above)
Comment (ppppbn) [reply webpack.test.js:20]: But the webpack config for testing is different with the common one. Like, a lot of different points.

File: src/linagora.esn.calendar/app/calendar/calendar.controller.spec.js (hunk 25 above)
Comment (tuanlc) [reply src/linagora.esn.calendar/app/calendar/calendar.controller.spec.js:12]: Not resolved yet

File: src/linagora.esn.calendar/app/components/calendars-list/items/item/calendars-list-item.controller.spec.js [hunk 41]
```diff
@@ -1,5 +1,7 @@
 'use strict';
 
+const { done } = require("q");
```
Comment (tuanlc) [inline src/linagora.esn.calendar/app/components/calendars-list/items/item/calendars-list-item.controller.spec.js:3]: Why??

File: src/linagora.esn.calendar/app/freebusy/confirmation-modal/event-freebusy-confirmation-modal.less [hunk 42]
```diff
@@ -1,7 +0,0 @@
-.event-freebusy-confirmation {
```
Comment (tuanlc) [inline src/linagora.esn.calendar/app/freebusy/confirmation-modal/event-freebusy-confirmation-modal.less:1]: The removing `free-busy` stuff should belong to a dedicated commit

File: src/linagora.esn.calendar/app/run.js [hunk 43]
```diff
@@ -1,18 +1,16 @@
...
-    $templateCache.put('/calendar/app/planning/aside/calendar-planning-for-aside.html', require('./planning/aside/calendar-planning-for-aside.pug'));
-    $templateCache.put('/calendar/app/sidebar/mobile/sidebar-mobile.html', require('./sidebar/mobile/sidebar-mobile.pug'));
-  }
-})();
+function addTemplateCache($templateCache) {
+  $templateCache.put('/calendar/app/search/event/event-search-item.html', require('./search/event/event-search-item.pug'));
+  $templateCache.put('/calendar/app/search/form/search-form-template.html', require('./search/form/search-form-template.pug'));
+  $templateCache.put('/calendar/app/planning/aside/calendar-planning-for-aside.html', require('./planning/aside/calendar-planning-for-aside.pug'));
+  $templateCache.put('/calendar/app/sidebar/mobile/sidebar-mobile.html', require('./sidebar/mobile/sidebar-mobile.pug'));
+}
```
Comment (tuanlc) [inline src/linagora.esn.calendar/app/run.js:16]: add blank line at the end

File: test/config/mocks/i18n-loader.js [hunk 44]
```diff
@@ -0,0 +1,26 @@
...
+          }
+
+          return catalogs[options.key];
+        });
+    };
+
+    function getCatalogs() {
+      return Promise.resolve({});
+    }
+  }
```
Comment (tuanlc) [inline test/config/mocks/i18n-loader.js:26]: Wrong indent

File: webpack.test.js (hunk 16 above)
Comment (huy-ta) [reply webpack.test.js]: No we shouldn't need babel for tests.

File: test/config/karma.conf.js [hunk 45]
```diff
@@ -0,0 +1,96 @@
...
+    webpack: webpackConfig,
+    plugins: [
+      'karma-chrome-launcher',
+      'karma-firefox-launcher',
+      'karma-mocha',
+      'karma-webpack',
+      'karma-coverage',
+      'karma-spec-reporter',
+      'karma-sinon-chai',
+      '@linagora/karma-ng-jade2module-preprocessor',
```
Comment (huy-ta) [inline test/config/karma.conf.js:68]: This is not needed because of webpack.

File: test/config/karma.conf.js [hunk 46]
```diff
@@ -0,0 +1,96 @@
...
+      'karma-rawfixtures-preprocessor'
+    ],
+
+    coverageReporter: {type: 'text', dir: '/tmp'},
+
+    ngJade2ModulePreprocessor: {
```
Comment (huy-ta) [inline test/config/karma.conf.js:74]: This is not needed because of webpack.

File: package.json [hunk 47]
```diff
@@ -22,8 +23,13 @@
     "@babel/core": "^7.10.5",
     "@babel/plugin-syntax-dynamic-import": "^7.8.3",
     "@babel/preset-env": "^7.10.4",
+    "@linagora/karma-ng-jade2module-preprocessor": "0.5.3",
```
Comment (huy-ta) [inline package.json:26]: This is not needed because of webpack.

File: src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js (hunk 21 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/services/calendar-attendee-service.spec.js]: Done :trollface:

File: src/esn.calendar.libs/app/services/fullcalendar/calendar-configuration.service.spec.js (hunk 39 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/services/fullcalendar/calendar-configuration.service.spec.js:8]: Sorry, I have removed this

File: src/esn.calendar.libs/app/services/shells/calendar-right-shell.spec.js (hunk 40 above)
Comment (ppppbn) [reply src/esn.calendar.libs/app/services/shells/calendar-right-shell.spec.js]: In a desperate moment :trollface: Have reverted it. Thanks!

File: src/linagora.esn.calendar/app/components/calendars-list/items/item/calendars-list-item.controller.spec.js (hunk 41 above)
Comment (ppppbn) [reply src/linagora.esn.calendar/app/components/calendars-list/items/item/calendars-list-item.controller.spec.js]: Sorry, VSCode auto import.

File: test/config/karma.conf.js (hunk 46 above)
Comment (ppppbn) [reply test/config/karma.conf.js:74]: We can only remove the `cacheIdfromPath` function. Others are still in need because we need sth to render our pug file. Try remove them and you will catch some errors i think.

File: src/linagora.esn.calendar/app/settings/calendars/settings-calendars.controller.spec.js [hunk 48]
```diff
@@ -82,11 +82,12 @@ describe('The CalSettingsCalendarsController controller', function() {
     it('should get the calendars from the calendarService', function() {
       var calendars = [calendar, otherCalendar];
 
-      CalSettingsCalendarsController.$onInit();
-      $rootScope.$digest();
+      CalSettingsCalendarsController.$onInit().then(function() {
```
Comment (piavgh) [inline src/linagora.esn.calendar/app/settings/calendars/settings-calendars.controller.spec.js:85]: This is async but no `done` is called?

Comment (ppppbn) [reply src/linagora.esn.calendar/app/settings/calendars/settings-calendars.controller.spec.js:85]: Nice catch. Fixed! Thank you.

**Comment by huy-ta** [comment]
I'll continue to work on this from now on.
